Hotfix/engine polish

jmidt opened the pull request:
Many minor fixes. Here goes:

Voicemanager: 
- on_note_on now passes an argument with the frequency target, if it is not enough to know the current frequency when triggering a note.
- Should now work for an arbitrary number of voices between 3 and 12. Not tested, though.

Chorus:
- Completely redesigned. Is now a tri-phase chorus instead of a quadrature chorus. Should be more lush. Also has possibilities for more flanger-like and resonator-filter-like settings.
- One remaining naggle is that there can be audio pops when changing parameters. However, changing parameters live is not really a typical chorus thing.

Example engines:
- Added an example synth and effect. These should probably be moved out and into a generating script, but for testing they are with the normal engines.

Master engine:
- Added softclip (tanh) on the output. Not exactly the smoothest of overdrives, but better than hardclipping

Arp:
- Fixed minor bugs with drawing the dots

Goss:
- Minor changes with the percussion (yellow) encoder. Should also be less clicky on retriggers now.

Rhodes:
- Much less clicky on retriggers now.
- I tried to make it work better with portamento, but without a complete redesign, it seems impossible.

New utils:
- Delayed lambda - for making stuff happen in the DSP later, some time after a trigger. Note it uses a virtual call so don't use it for repeated stuff.
- BiquadSoftReset. A biquad class with the possibility of freezing the output when you change its inner workings. This can avoid nasty audio clicks. A more elegant solution for many (but not all) use cases, is the Direct Form I issue opened recently.

topisani commented on src/core/voices/voice_manager.hpp:
```diff
@@ -272,8 +274,10 @@ namespace otto::core::voices {
-    std::array<float, 6> rand_max = {0.94, 0.999, 1.03, 1.06, 0.92, 1.01};
+    // Random values. 100% random, organic and fresh. Works for up to 12 voices.
+    std::vector<float> rand_max_length = {0.94, 0.999, 1.03, 1.06, 0.92, 1.01,
+                                                  1.02, 0.98, 1.0, 1.09, 0.94, 1.05};
```
Nope, get those back in a fixed size array of size NumberOfVoices, and don't use the rand_max_length member just for initializing the second one. You can initialize it using a lambda if you want.

jmidt replied:
I am a bit unsure how to do this. I want it to be the same numbers every time, so it is somewhat predictable. Could you give an example?

jmidt replied:
It is now just an array with 12 elements. Inelegant, but will work as long as the voice count is not above that.

jmidt commented:
Should I rebase it on develop first?

jmidt commented:
- Changed soft clipping to arctan. A bit smoother.
- Fixed bugs in the voice manager:
    - Voices are now reused on repeated triggers
    - `on_note_off` is note sent on note steals if legato is engaged.
- Rhodes sounds better (more bass)
- Potion sounds better (less clicks)
- Synths have been adjusted to roughly the same output level.

jmidt commented:
Should be ready to merge into develop if travis agrees.
